changeset/manager.go

```diff
@@ -0,0 +1,432 @@
+	errChangeTypeNotSet = errors.New("changesType must be specified")
+)
+
+// ManagerOptions are options used in creating a new ChangeSet Manager
+type ManagerOptions interface {
+	// KV is the KVStore holding the configuration
+	KV(kv kv.Store) ManagerOptions
+	GetKV() kv.Store
+
+	// ConfigKey is the key holding the configuration object
+	ConfigKey(key string) ManagerOptions
+	GetConfigKey() string
+
+	// Logger is the logger to use
+	Logger(logger xlog.Logger) ManagerOptions
+	GetLogger() xlog.Logger
+
+	// ConfigType is a proto.Message defining the structure of the configuration
+	// object.  Clones of this proto will be used to unmarshal configuration
+	// instances
+	ConfigType(config proto.Message) ManagerOptions
+	GetConfigType() proto.Message
+
+	// ChangesType is a proto.Message defining the structure of the changes
+	// object.  Clones of this protol will be used to unmarshal change list
+	// instances.
+	ChangesType(changes proto.Message) ManagerOptions
+	GetChangesType() proto.Message
+
+	// Validate validates the options
+	Validate() error
+}
+
+// NewManagerOptions creates an empty ManagerOptions
+func NewManagerOptions() ManagerOptions { return new(managerOptions) }
+
+// A ChangeFn adds a change to an existing set of changes
+type ChangeFn func(config, changes proto.Message) error
+
+// An ApplyFn applies a set of changes to a configuration, resulting in a new
+// configuration
+type ApplyFn func(config, changes proto.Message) error
+
+// A Manager manages sets of changes in a version friendly manager.  Changes to
+// a given version of a configuration object are stored under
+// <key>/_changes/<version>.  Multiple changes can be added, then committed all
+// at once.  Committing transforms the configuration according to the changes,
+// then writes the configuration back.  CAS operations are used to ensure that
+// commits are not applied more than once, and to avoid conflicts on the change
+// object itself.
+type Manager interface {
+	// Change creates a new change against the latest configuration, adding it
+	// to the set of pending changes for that configuration
+	Change(change ChangeFn) error
+
+	// GetPendingChanges gets the latest uncommitted changes
+	GetPendingChanges() (int, proto.Message, proto.Message, error)
+
+	// Commit commits the specified ChangeSet, transforming the configuration on
+	// which they are based into a new configuration, and storing that new
+	// configuration as a next versions. Ensures that changes are applied as a
+	// batch, are not applied more than once, and that new changes are not
+	// started while a commit is underway
+	Commit(version int, apply ApplyFn) error
+}
+
+// NewManager creates a new change list Manager
+func NewManager(opts ManagerOptions) (Manager, error) {
+	if opts == nil {
+		return nil, errOptsNotSet
+	}
+
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	logger := opts.GetLogger()
+	if logger == nil {
+		logger = xlog.NullLogger
+	}
+
+	return &manager{
```

Comment:
Since the member methods are all defined on the value and the manager instance is essentially immutable, could probably return a value here to avoid a heap allocation?